CPU scheduling for workloads

serxa opened the pull request:
This PR implements concurrency control on top of the resource scheduler and introduces a new type of RESOURCE and a new WORKLOAD setting to limit the number of concurrent threads for a specific workload.

Usage example:
```
CREATE RESOURCE cpu (MASTER THREAD, WORKER THREAD)
CREATE WORKLOAD all
CREATE WORKLOAD admin IN all SETTINGS max_concurrent_threads = 10
CREATE WORKLOAD production IN all SETTINGS max_concurrent_threads = 100
CREATE WORKLOAD analytics IN production SETTINGS max_concurrent_threads = 60, weight = 9
CREATE WORKLOAD ingestion IN production
```

This configuration example provides independent CPU slot pools for admin and production. The production pool is shared between analytics and ingestion. Furthermore, if the production pool is overloaded, 9 of 10 released slots will be rescheduled to analytical queries if necessary. The ingestion queries would only receive 1 of 10 slots during overload periods. This might improve the latency of user-facing queries. Analytics has its own limit of 60 concurrent thread, always leaving at least 40 threads to support ingestion. When there is no overload, ingestion could use all 100 threads.

WARNING: Slot scheduling provides a way to control [query concurrency](/operations/settings/settings.md/#max_threads) but does not guarantee fair CPU time allocation yet. This requires further development of CPU slot preemption and will be supported in the follow-up PR.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added CPU slot scheduling for workloads, see https://clickhouse.com/docs/operations/workload-scheduling#cpu_scheduling for details

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->

serxa committed:
support CPU keyword in RESOURCE description

serxa committed:
Add cpu scheduler

serxa committed:
integrate cpu scheduler with PipelineExecutor

serxa committed:
add unit test for cpu slots scheduler

serxa committed:
fix test

serxa committed:
fix hangup in unittest

serxa committed:
Use separate scheduling settings for CPU resource constraints

serxa committed:
udpate docs for cpu and io scheduling

serxa committed:
Merge branch 'master' into cpu-scheduler

serxa committed:
rework CPU access mode into two MASTER THREAD and WORKER THREAD

serxa committed:
Merge branch 'cpu-scheduler' of github.com:ClickHouse/ClickHouse into cpu-scheduler

serxa committed:
fix style

serxa committed:
fix style 2

serxa committed:
more docs

serxa committed:
fix docs build

bharatnc commented on src/Parsers/ASTCreateResourceQuery.h:
```diff
@@ -10,15 +12,49 @@ namespace DB
+        CPUSlot,
+    };
+
+    /// Describes kinds of operations requiring the access to the resourece
```
```suggestion
    /// Describes kinds of operations requiring the access to the resource
```

serxa committed:
Update ASTCreateResourceQuery.h

Co-authored-by: Bharat Nallan <bharatnc@gmail.com>

serxa committed:
fix scheduling settings updating

serxa committed:
Merge branch 'cpu-scheduler' of github.com:ClickHouse/ClickHouse into cpu-scheduler

serxa committed:
fix deadlock

serxa committed:
add scheduler thread names

serxa committed:
add cpu slot fair allocation test

serxa committed:
add test with independent cpu slot pools

serxa committed:
add test for cpu scheduling weights

serxa committed:
add test for cpu scheduling with priorities

serxa committed:
fix integration test

serxa committed:
fix cpu slots allocation for unknown workload

serxa committed:
fix hanging cpu slots

serxa committed:
add cpu scheduling integration test

serxa committed:
reproduce request cancelation bug in unittest

serxa commented:
It seems that not all scheduling nodes were modified to support resource request cancelling in https://github.com/ClickHouse/ClickHouse/pull/60558. SemaphoreConstraint and ThrottlerConstraint were omitted. Resource request canceling is not used with IO scheduling, so this bug was not noticed until now. Fixing...

serxa committed:
support request canceling in all scheduling nodes

serxa committed:
more backward-compatibility testing

serxa committed:
add more integration tests for cpu scheduling

serxa committed:
fix docs build

serxa committed:
bugfix: truncate scheduler thread name

serxa committed:
fix create resource query formating

serxa committed:
rename: SchedulingSettings -> WorkloadSettings

serxa committed:
thrown on unknown workload setting

serxa committed:
support max_concurrent_threads_ratio_to_cores workload setting

serxa committed:
bugfix: never pass zero limit to semaphore constraint

serxa committed:
add current metric and profile event for scheduled slot requests

serxa commented:
I've added the tests that cover all main features and added metrics similar to those we had for IO scheduling and concurrency control. I've made a few renamings for clarity: SchedulingSettings -> WorkloadSettings, IOResourceManager -> WorkloadResourceManager. And WORKLOAD settings were also renamed, because we now have settings for both IO and CPU, but backward compatibility for old settings names is kept.

@alexkats This PR is now ready to be reviewed in full. I hope it will survive CI. Anyway, no code changes are expected on my side, besides those necessary to make CI green.

serxa committed:
Merge branch 'master' into cpu-scheduler

serxa committed:
fix darwin build

serxa committed:
try fix docs build

serxa committed:
fix stack-use-after-return in unit test

serxa commented:
> [Stateless tests (tsan, s3 storage, 3/3)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=77595&sha=27a349f81be8a4ffbf73b8d1d1483310974358cb&name_0=PR&name_1=Stateless%20tests%20%28tsan%2C%20s3%20storage%2C%203%2F3%29) — Server died, fail: 13, passed: 692, skipped: 13

https://github.com/ClickHouse/ClickHouse/issues/77600

serxa commented:
> [Stateless tests (debug, s3 storage)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=77595&sha=27a349f81be8a4ffbf73b8d1d1483310974358cb&name_0=PR&name_1=Stateless%20tests%20%28debug%2C%20s3%20storage%29) — fail: 1, passed: 7459, skipped: 326

https://github.com/ClickHouse/ClickHouse/issues/78366

serxa committed:
Merge branch 'master' into cpu-scheduler

alexkats commented on src/Common/Scheduler/CPUSlotsAllocation.h:
```diff
@@ -0,0 +1,101 @@
+    void grant();
+
+    // Returns the queue for the current request
+    ISchedulerQueue * getCurrentQueue(const std::unique_lock<std::mutex> &);
```
```suggestion
    ISchedulerQueue * getCurrentQueue(const std::unique_lock<std::mutex> &) const;
```

alexkats commented on src/Common/Scheduler/CPUSlotsAllocation.cpp:
```diff
@@ -0,0 +1,222 @@
+    wait_timer.reset();
+}
+
+ISchedulerQueue * CPUSlotsAllocation::getCurrentQueue(const std::unique_lock<std::mutex> &)
```
```suggestion
ISchedulerQueue * CPUSlotsAllocation::getCurrentQueue(const std::unique_lock<std::mutex> &) const
```

alexkats commented on tests/queries/0_stateless/03232_workloads_and_resources.sql:
```diff
@@ -19,13 +19,14 @@ create workload invalid in all settings priority = 'invalid_value'; -- {serverEr
+create workload invalid in all settings max_bytes_inflight = -1; -- {serverError BAD_ARGUMENTS}
+create workload invalid in all settings unknown_setting = 42; -- {serverError BAD_ARGUMENTS}
+create workload invalid in all settings max_io_requests = -1; -- {serverError BAD_ARGUMENTS}
+create workload invalid in all settings max_io_requests = 1.5; -- {serverError BAD_GET}
```
Maybe add some cases with mixed resources as well?

serxa replied:
Yes, good idea

serxa replied:
Added

serxa commented on src/Common/Scheduler/Nodes/UnifiedSchedulerNode.h:
```diff
@@ -567,15 +567,25 @@ class UnifiedSchedulerNode final : public ISchedulerNode
-            busy_periods++;
-        incrementDequeued(request->cost);
-        return {request, child_active};
+        // Cycle is required to do deactivations in the case of canceled requests, when dequeueRequest returns `nullptr`
+        while (true)
+        {
+            // Dequeue request from the child
+            auto [request, child_now_active] = immediate_child->dequeueRequest();
+
+            // Deactivate if necessary
+            child_active = child_now_active;
+            if (!child_active)
+                busy_periods++;
+
+            if (request)
+            {
+                incrementDequeued(request->cost);
+                return {request, child_active};
+            }
+            else
+                return {nullptr, false};
+        }
```
Yes, good catch. I copy-pasted it from FairPolicy and reworked a bit to get this nonsense. No need for cycle here. FairPolicy could have multiple children, so if one deactivated due to cancel we should go to another one, but UnifiedSchedulerNode, SemaphoreConstraint, and ThrottlerConstraint can have the only immediate child.

serxa committed:
remove non-sense cycles

serxa commented on src/Common/Scheduler/Nodes/CustomResourceManager.cpp:
```diff
@@ -180,7 +180,7 @@ ResourceLink CustomResourceManager::Classifier::get(const String & resource_name
     : state(new State())
 {
-    scheduler.start();
+    scheduler.start("Sch.CstmResMngr");
```
There are only 15 characters for the thread names, unfortunately

alexkats replied:
Oh, I didn't know, ok then

serxa commented on src/Common/Scheduler/Nodes/SemaphoreConstraint.h:
It is used by the `SchedulerNodeFactory` and it should follow the common signature as all other nodes types to be compilable, see `SchedulerNodeFactory::registerMethod()` template. I hope we will get rid of this deprecated XML configuration for the scheduler and remove all these node constructors at some point.

serxa commented on src/Common/Scheduler/Nodes/WorkloadResourceManager.cpp:
```diff
@@ -38,29 +37,40 @@ namespace
+    , unit(getResourceUnit(resource_entity_))
 {
-    scheduler.start();
+    scheduler.start("Sch." + resource_name);
```
Again 15 letters :-(

serxa commented on src/Common/Scheduler/CPUSlotsAllocation.cpp:
```diff
@@ -0,0 +1,222 @@
+    wait_timer.reset();
+}
+
+ISchedulerQueue * CPUSlotsAllocation::getCurrentQueue(const std::unique_lock<std::mutex> &)
+{
+    return allocated < master_slots ? master_link.queue : worker_link.queue;
+}
```
Yes, right. I'll add the comment.

serxa committed:
add comments

serxa committed:
Update docs/en/operations/workload-scheduling.md

Co-authored-by: Alexey Katsman <alex.katsman@clickhouse.com>

serxa committed:
Update src/Common/Scheduler/CPUSlotsAllocation.h

Co-authored-by: Alexey Katsman <alex.katsman@clickhouse.com>

serxa committed:
Update src/Common/Scheduler/CPUSlotsAllocation.cpp

Co-authored-by: Alexey Katsman <alex.katsman@clickhouse.com>

serxa committed:
const-correctness

serxa committed:
Merge branch 'cpu-scheduler' of github.com:ClickHouse/ClickHouse into cpu-scheduler

serxa committed:
Update docs/en/operations/workload-scheduling.md

Co-authored-by: Alexey Katsman <alex.katsman@clickhouse.com>

serxa committed:
fix mutex

serxa committed:
Update docs/en/operations/workload-scheduling.md

Co-authored-by: Alexey Katsman <alex.katsman@clickhouse.com>

serxa committed:
test cpu and IO mixed resource

serxa committed:
Merge branch 'cpu-scheduler' of github.com:ClickHouse/ClickHouse into cpu-scheduler

serxa committed:
add more comments

alexkats reviewed (approved):
LGTM, thanks

clickgapai commented:
Hi @serxa @alexkats — while reviewing this PR I found the following:

- **Bug (P2):** getNotNegativeFloat64 accepts negative Float64 values for workload settings — https://github.com/ClickHouse/ClickHouse/issues/101825

Happy to discuss — close anything that's wrong or already addressed.
